source/extensions/filters/network/zookeeper_proxy/zookeeper_decoder.cc

```diff
@@ -0,0 +1,344 @@
+                                          const bool two) {
+  checkLength(len, 14);
+
+  std::string path = BufferHelper::peekString(data, offset);
+  bool watch = BufferHelper::peekBool(data, offset);
+
+  callbacks_.onGetChildrenRequest(path, watch, two);
+}
+
+void DecoderImpl::parseDeleteRequest(Buffer::Instance& data, uint64_t& offset, uint32_t len) {
+  checkLength(len, 16);
+
+  std::string path = BufferHelper::peekString(data, offset);
+  int32_t version = BufferHelper::peekInt32(data, offset);
+
+  callbacks_.onDeleteRequest(path, version);
+}
+
+void DecoderImpl::parseExistsRequest(Buffer::Instance& data, uint64_t& offset, uint32_t len) {
+  checkLength(len, 13);
+
+  std::string path = BufferHelper::peekString(data, offset);
+  bool watch = BufferHelper::peekBool(data, offset);
+
+  callbacks_.onExistsRequest(path, watch);
+}
+
+void DecoderImpl::parseGetAclRequest(Buffer::Instance& data, uint64_t& offset, uint32_t len) {
+  checkLength(len, 8);
+
+  std::string path = BufferHelper::peekString(data, offset);
+
+  callbacks_.onGetAclRequest(path);
+}
+
+void DecoderImpl::parseSetAclRequest(Buffer::Instance& data, uint64_t& offset, uint32_t len) {
+  checkLength(len, 8);
+
+  std::string path = BufferHelper::peekString(data, offset);
+  skipAcls(data, offset);
+  int32_t version = BufferHelper::peekInt32(data, offset);
+
+  callbacks_.onSetAclRequest(path, version);
+}
+
+void DecoderImpl::parseSyncRequest(Buffer::Instance& data, uint64_t& offset, uint32_t len) {
+  checkLength(len, 8);
+
+  std::string path = BufferHelper::peekString(data, offset);
+
+  callbacks_.onSyncRequest(path);
+}
+
+void DecoderImpl::parseCheckRequest(Buffer::Instance& data, uint64_t& offset, uint32_t len) {
+  checkLength(len, 8);
+
+  std::string path = BufferHelper::peekString(data, offset);
+  int32_t version = BufferHelper::peekInt32(data, offset);
+
+  callbacks_.onCheckRequest(path, version);
+}
+
+void DecoderImpl::parseMultiRequest(Buffer::Instance& data, uint64_t& offset, uint32_t len) {
+  // Treat empty transactions as a decoding error, there should be at least 1 header.
+  checkLength(len, 17);
+
+  while (true) {
+    int32_t type = BufferHelper::peekInt32(data, offset);
+    bool done = BufferHelper::peekBool(data, offset);
+    // Ignore error field.
+    BufferHelper::peekInt32(data, offset);
+
+    if (done) {
+      break;
+    }
+
+    switch (type) {
+    case enumToInt(OpCodes::CREATE):
+      parseCreateRequest(data, offset, len, false);
+      break;
+    case enumToInt(OpCodes::SETDATA):
+      parseSetRequest(data, offset, len);
+      break;
+    case enumToInt(OpCodes::CHECK):
+      parseCheckRequest(data, offset, len);
+      break;
+    default:
+      // Should not happen.
+      throw EnvoyException("Unknown type within a transaction");
+    }
+  }
+
+  callbacks_.onMultiRequest();
+}
+
+void DecoderImpl::parseReconfigRequest(Buffer::Instance& data, uint64_t& offset, uint32_t len) {
+  checkLength(len, 28);
```

Comment:
maybe this should be called something like `ensureMinLength` to make it clearer what it's doing?